src/components/fields/derived.js

```diff
@@ -461,15 +461,63 @@ function computeAxesRefOptions(axes, propsAttr) {
   return options;
 }
 
+export const TextPosition = connectToContainer(UnconnectedDropdown, {
+  modifyPlotProps: (props, context, plotProps) => {
+    const {localize: _} = props;
+    let options = [
+      {label: _('Top Left'), value: 'top left'},
+      {label: _('Top Center'), value: 'top center'},
+      {label: _('Top Right'), value: 'top right'},
+      {label: _('Middle Left'), value: 'middle left'},
+      {label: _('Middle Center'), value: 'middle center'},
+      {label: _('Middle Right'), value: 'middle right'},
+      {label: _('Bottom Left'), value: 'bottom left'},
+      {label: _('Bottom Center'), value: 'bottom center'},
+      {label: _('Bottom Right'), value: 'bottom right'},
+    ];
+    if (context.container.type === 'pie') {
+      options = [
+        {label: _('Inside'), value: 'inside'},
+        {label: _('Outside'), value: 'outside'},
+        {label: _('Auto'), value: 'auto'},
+        {label: _('None'), value: 'none'},
+      ];
+    }
+    plotProps.options = options;
+    plotProps.clearable = false;
+  },
+});
+
 export const HoverInfo = connectToContainer(UnconnectedFlaglist, {
   modifyPlotProps: (props, context, plotProps) => {
     const {localize: _} = props;
     let options = [
       {label: _('X'), value: 'x'},
       {label: _('Y'), value: 'y'},
-      {label: _('Z'), value: 'z'},
+      {label: _('Name'), value: 'name'},
     ];
 
+    if (
+      [
+        'heatmap',
+        'heatmapgl',
+        'histogram2d',
+        'histogram2dcontour',
+        'contour',
+        'contourcarpet',
+        'scatter3d',
+        'surface',
+        'mesh3d',
+      ].includes(context.container.type)
```

Comment:
can't all of these come from the Plotly Schema? Doesn't the schema hold a list of all the available options for each trace? Do we need to duplicate it here?

Reply:
it doesn't come from the schema, this is not a duplicate..

Reply:
Right, ideally we would pick out the `flags` from the schema for each trace type, push each into `options` here along with its translation, and then *maybe* filter based on what's actually provided in this trace - for example `text` should be available for scatter and bar traces (and it will be based on the `flags` in the schema), but it will have no effect if there's no `text` or `hovertext` field present in the trace so could be omitted.